crates/common/mqtt_channel/src/connection.rs

```diff
@@ -252,6 +265,11 @@ impl Connection {
                     break;
                 }
 
+                Ok(Event::Incoming(Incoming::PubAck(_))) => {
+                    // Mark one message as consumed
+                    awaiting_acks.next().await;
+                }
```

Comment:
We also have to wait for QoS 2 completion.

```suggestion
                Ok(Event::Incoming(Incoming::PubAck(_))) | Ok(Event::Incoming(Incoming::PubComp(_))) => {
                    // Mark one message as consumed
                    awaiting_acks.next().await;
                }
```